backend/src/controllers/api-key.controller.ts

```diff
@@ -0,0 +1,87 @@
+import { BaseController } from './base.controller';
+import {
+  IApiKey,
+  IApiKeyMinimal,
+  IApiKeyMinimalSchema,
+  ICreateApiKey,
+  ICreateApiKeyResponse,
+  ICreateApiKeyResponseSchema,
+  IRevokeApiKey,
+} from 'shared-types';
+import { Response } from 'express';
+import { randomBytes } from 'crypto';
+import { hash } from 'bcrypt';
+import ApiKey from '../models/api-key.model';
+import { AuthRequest } from '../common/auth.middleware';
+import { daysAhead } from '../utils/date';
+import httpStatus from 'http-status';
+
+class ApiKeyController extends BaseController<IApiKey> {
+  constructor() {
+    super(ApiKey);
+  }
+
+  async getAll(req: AuthRequest<object, IApiKeyMinimal[]>, res: Response<IApiKeyMinimal[]>) {
+    const { businessId } = req.user!;
+    this.debug(`Getting all API keys for business ${businessId}`);
+
+    const keys = await ApiKey.find({ businessId });
+    const sanitizedKeys = keys.map((keyData) => IApiKeyMinimalSchema.parse(keyData));
+
+    return res.status(httpStatus.OK).json(sanitizedKeys);
+  }
+
+  async generateApiKey(
+    req: AuthRequest<object, ICreateApiKeyResponse, ICreateApiKey>,
+    res: Response<ICreateApiKeyResponse>
+  ) {
+    const { businessId, email } = req.user!;
+    const expiry = new Date(req.body.expiry || daysAhead(365));
+
+    this.debug(`Generating API key for ${businessId} as requested by user mail "${email}"`);
+
+    try {
+      const plainApiKey = `${businessId}:${randomBytes(32).toString('hex')}`;
+      const hashedApiKey = await hash(plainApiKey, 10);
+
+      const newApiKey = new ApiKey({ key: hashedApiKey, businessId, expiry, revoked: false });
+      await newApiKey.save();
+      return res.status(httpStatus.CREATED).json(ICreateApiKeyResponseSchema.parse({ ...newApiKey, key: plainApiKey }));
+    } catch (err) {
+      this.debug(
+        `Error generating API key for ${businessId} as requested by user mail "${email}"`,
+        (err as Error).message,
+        (err as Error).stack || 'no stacktrace'
+      );
+      return res.status(httpStatus.INTERNAL_SERVER_ERROR).send();
+    }
+  }
+
+  async revokeKey(req: AuthRequest<IRevokeApiKey, IApiKey>, res: Response) {
```

Comment:
Shouldnt we check that the api key entered in the request is owned by the user that send it?

Reply:
Look at the query in line 67. I am querying a key that has the `businessId` from the user's headers and the `id` he provided to me.
He has to own it otherwise he wouldn't have had the JWT token.